src/gh-api-client.ts

```diff
@@ -12,20 +13,39 @@ import { validateObject } from "./json-validation";
 
 export const userAgent = "GitHub multi-repository variant analysis action";
 
-function getOctokitRequestInterface(): RequestInterface {
-  const octokit = new Octokit({ userAgent, retry });
-
-  const signedAuthToken = getSignedAuthToken();
-  return octokit.request.defaults({
-    request: {
-      hook: (request: RequestInterface, options: EndpointOptions) => {
-        if (options.headers) {
-          options.headers.authorization = `RemoteAuth ${signedAuthToken}`;
-        }
-        return request(options);
+function getOctokit(): Octokit {
+  const throttlingOctokit = Octokit.plugin(throttling);
+  const octokit = new throttlingOctokit({
+    userAgent,
+    retry,
+    authStrategy: () => {
+      return {
+        hook: (request: RequestInterface, options: EndpointOptions) => {
+          if (options.headers) {
+            options.headers.authorization = `RemoteAuth ${getSignedAuthToken()}`;
+          }
+          return request(options);
+        },
+      };
+    },
+    throttle: {
+      enabled: !!process.env.CODEQL_VARIANT_ANALYSIS_ACTION_WAIT_ON_RATE_LIMIT,
```

Comment:
Does this correctly handle a "false" value?

I believe `process.env.CODEQL_VARIANT_ANALYSIS_ACTION_WAIT_ON_RATE_LIMIT` will always be a string value, but the value may be the strings "true" or "false". So then using `!!` on this will always yield `true`, because javascript just looks at whether the string is non-empty or not.

That said, do we need this functionality to be gated behind a feature flag or is it something we should enable for all users? I can't see any downsides of it or why a user would want to turn it off. 🤷🏼

Reply:
That's a great catch! IIRC from the testing I did, the absence of the variable repository made `enabled: false`, but I agree it's not ideal.

> do we need this functionality to be gated behind a feature flag

I'm guessing we do, as private repositories would get their Actions usage increased a lot ($$$) in case of a rate limit triggered, but I'm happy to remove the "gate".

Reply:
> That makes sense, but is the variable ever absent, because we do ${{ (vars.MRVA_WAIT_ON_RATE_LIMIT && true) || false }}? In the false case what value does the env var get inside nodejs?

Oops, I just re-tried and it works as you say, so https://github.com/github/codeql-variant-analysis-action/pull/974/commits/162ff3bdf1179e31ee868fc83d71f8506e57d630 makes the default value a blank string to fix the discrepancy.

> Would it make sense to add a limit so it only retries a bounded number of times? That way it wouldn't increase costs by very much.

That would work, but still increase the use a lot and not work for when the rate limit is triggered more than once at a time.

> Out of interest, who are the intended users right now? (You're welcome to message us on slack or in an internal issue if you don't want to say here on this public PR).

Me! We are doing heavy use of MRVA lately and this helps me launch queries without worrying about them being canceled.